### PR TITLE
Kubeadm graduate addons phase

### PR DESCRIPTION
**Pull request by yagonobre**
**What type of PR is this?**
This PR graduates phase addon in the kubeadm init workflow.
/kind feature

**What this PR does / why we need it**:
Ref https://github.com/kubernetes/kubeadm/issues/1163

/sig cluster-lifecycle
/cc @neolit123
@kubernetes/sig-cluster-lifecycle-pr-reviews

```release-note
NONE
```

### CODE REVIEW

File: cmd/kubeadm/app/cmd/phases/addons.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/addons.go]: i don't think we would need aliases for the addon phase.

File: cmd/kubeadm/app/cmd/init.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/init.go]: hm, why the rename?

File: cmd/kubeadm/app/cmd/phases/addons.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/addons.go]: "all" is not needed because the phase runner will run all subphases unless a sub phase is specified.

File: cmd/kubeadm/app/cmd/init.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/init.go]: @fabriziopandini this forces us to have a kubeconfig on the init CLI. :thinking:

File: cmd/kubeadm/app/cmd/init.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/init.go]: i think we should remove this flag. the image repository is already exposed via the config.
but also we have some new image related fields for the addons in 1.13.

File: cmd/kubeadm/app/cmd/phases/addons.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/addons.go]: return err.

File: cmd/kubeadm/app/cmd/phases/addons.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/addons.go]: return err.

File: cmd/kubeadm/app/cmd/phases/addons.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/addons.go]: since `all` is removed we can remove this function too.

File: cmd/kubeadm/app/cmd/phases/addons.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/addons.go]: we tend to omit empty new lines like this one. :)

**Review by neolit123 — COMMENTED** [review]
@yagonobre 
thanks for the PR!
i've added some initial feedback.

File: cmd/kubeadm/app/cmd/phases/addons.go
Comment (yagonobre) [reply cmd/kubeadm/app/cmd/phases/addons.go]: nice!

File: cmd/kubeadm/app/cmd/init.go
Comment (yagonobre) [reply cmd/kubeadm/app/cmd/init.go]: I imported a `options` pkg

**Comment by yagonobre** [comment]
@neolit123 updated, also I'm removing the wip

**Comment by yagonobre** [comment]
rebased

**Comment by timothysc** [comment]
@yagonobre - please rebase and squash to no more then 2 commits on any given PR.

**Comment by yagonobre** [comment]
@timothysc done
/hold

**Review by timothysc — APPROVED** [review]
/approve 

still hold.  
It seems like you're missing tests now.

**Comment by neolit123** [comment]
a side effect of the `Phase struct` transition is that we have to discard the unit tests previously running cobra commands directly. e.g. addons_test.go

@fabriziopandini i wonder how should approach this :thinking: custom flags per phase and custom tests per phase seem to be related.

**Review by timothysc — APPROVED** [review]
/approve 

per SIG call.

**Comment by timothysc** [comment]
/hold cancel 

@fabriziopandini for final LGTM

File: cmd/kubeadm/app/cmd/phases/addons.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/addons.go]: we should remove Long.
it would contain:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/cmd/util/documentation.go#L31

it's not true because the 'addon' command can be called on it's own (calls both sub-phases).

File: cmd/kubeadm/app/cmd/phases/addons.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/addons.go]: please rename the var to coreDNSAddonLongDesc

File: cmd/kubeadm/app/cmd/phases/addons.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/addons.go]: please rename to kubeProxyAddonLongDesc

**Review by neolit123 — COMMENTED** [review]
@yagonobre thank you.
added a couple of final, small comments.

File: cmd/kubeadm/app/cmd/phases/addons.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/phases/addons.go]: not blocking.
I think that we can remove ensure (same for runEnsureKubeProxyAddon below)

File: cmd/kubeadm/app/cmd/phases/addons.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/phases/addons.go]: it seems that some flag are missing:
- kubernetes-version
- image respository (that if I'm not wrong should be added to init [here](https://github.com/kubernetes/kubernetes/blob/b8fece50f533bdfbf58b638f73e18a5773c3434c/cmd/kubeadm/app/cmd/init.go#L168-L172) by creating `options.AddImageMetaFlags`)

File: cmd/kubeadm/app/cmd/phases/addons.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/phases/addons.go]: I'm fine in removing this considering kubernetes/kubernetes: Pull Request 70024 will merge soon.
However, once this PR merges probably it will be necessary to add a new DNSType flag to allow users to choose between CoreDNS and kube-DNS.

@timothysc @neolit123  opinions?

**Review by fabriziopandini — APPROVED** [review]
@yagonobre well done!
only one small thing to be addressed with regards flags and then ready to go!
/approve

@neolit123 
> custom flags per phase and custom tests per phase seem to be related issues.

I don't think we have an issue, and IMO now there are simpler and better way to test flags (by testing getPhaseflags func) or phase logic (by testing runPhase func without having to deal with UX).

Nevertheless, if you want to go for the hard way and test things passing through cobra cmd, this can still be achieved by passing through `kubeadm init`, but IMO this should be used only for few integration tests.

File: cmd/kubeadm/app/cmd/phases/addons.go
Comment (neolit123) [reply cmd/kubeadm/app/cmd/phases/addons.go]: adding the CLI flag to chose DNS could be nice to have, then again we can avoid it.
in the discussions in this PR: https://github.com/kubernetes/website/pull/10294#discussion_r219645906
we discussed to drop GA feature gates that are doing toggling (like our CoreDNS gate) and to have a flag, instead.

then again, given with have it in the configuration the flag is not really needed, so my vote would be to only be able to pick DNS from the config.

**Comment by neolit123** [comment]
@fabriziopandini 
> I don't think we have an issue, and IMO now there are simpler and better way to test flags (by testing getPhaseflags func) or phase logic (by testing runPhase func without having to deal with UX).

i agree that this drastically simplifies our testing.
i've also expressed that i like having all init / phase flags in one place even if some of the flags are unrelated to init directly.

File: cmd/kubeadm/app/cmd/phases/addons.go
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/addons.go]: we still need to have image repository here?

Comment (yagonobre) [reply cmd/kubeadm/app/cmd/phases/addons.go]: yes, updated!

Comment (neolit123) [reply cmd/kubeadm/app/cmd/phases/addons.go]: thanks, waiting on CI and we can merge.

**Review by fabriziopandini — APPROVED** [review]
@yagonobre well done!
/lgtm
/approve

**Comment by neolit123** [comment]
awesome, thanks a lot @yagonobre 
waiting on CI and LGTM-ing.

**Pull request merged**
